0000-qiskit-pulse-schedule-relative-timing.md

```diff
@@ -52,49 +52,107 @@ Factors to consider:
 |---------------|----------------|------------------|
-|`barrier(self)`| O(\|C\|) | Add `Delay`s across `channels` so that all of the `channels` are occupied until `self.duration(channels)`. |
-|`left_align(self)` | O(\|C\|) | This is the default behavior. We could have it return `self`, or we could remove any pre-pended `Delay`s. The latter is probably better, because `schedule.right_align().left_align()` should yield the original `schedule`. Should we also append `Delay`s? Likely. *Needs more thought* |
-|`center_align(self)` | O(\|C\|) | Pre- and ap-pend equally sized `Delay`s (how to split odd durations?) on each channel so that all channels are occupied until `self.duration()` |
-|`right_align(self)` | O(\|C\|) | Prepend `Delay`s on each channel so that all channels are occupied until `self.duration()`. Should we also remove trailing `Delay`s? |
-
+|`barrier(self, channels=None)`| O(\|C\|) | Add `Delay`s across `channels` so that all of the `channels` are occupied until `self.duration(channels)`. By default, barrier on all channels. |
+|`left_align(self)` | O(\|C\|) | This is the default behavior. Return `self`.
+|`center_align(self)` | O(\|C\|) | Pre- and append equally sized `Delay`s (how to split odd durations?) on each channel so that all channels are occupied until `self.duration()` |
+|`right_align(self)` | O(\|C\|) | Prepend `Delay`s on each channel so that all channels are occupied until `self.duration()`.
+
+**Example**
+
+```
+sched = Schedule()
+sched += Delay(10, drive0)
+sched += Play(gaussian, drive0)
+sched = sched.barrier(drive0, drive1)
+subsched = Play([1.0]*20, drive1)
+subsched = Play([1.0]*10, drive0)
+sched += subsched.right_align()
+```
 
 ### API Changes
 
-\# TODO
-
 | Method        | Runtime        |  Change          |
 |---------------|----------------|------------------|
-| `append(self, schedule)` | O(\|N<sub>new</sub>\|) |  |
-|`insert(self, time, schedule)`| O(\|N<sub>new</sub>\| * \|N<sub>C</sub>\|) |  |
+| `append(self, schedule)` | O(\|C\|) | Previously, `append` would add `schedule` to `self` with relative times preserved, at time set by `self.duration({schedule.channels}.intersection({self.channels}))`. The new method will simply extend the instructions in `self` with those in `schedule` on a per channel basis. |
+|`insert(self, time, schedule)`| O(N<sub>new</sub> * N<sub>C</sub>) | This will be slow generally, but it can be implemented with the same runtime as `append` when the `time` is greater than `self.duration(schedule.channels)`. |
+
+**Example**
+
+```
+# To recreate previous append behavior:
+sched = sched.barrier(sched2.channels)
+sched += sched2
+```
 
 ## Detailed Design
 
-\# TODO
- - How internal data will now be tracked
- - How new and existing methods will be updated
+The `Schedule` will have an internal attribute, `_insts_by_chan`:
+
+```
+_insts_by_chan: Dict[Channel, List[ScheduleNode]]
+```
+
+**Example**
+```
+# sched._insts_by_chan
```

Comment:
This doesn't necessarily work as many instructions are on multiple channels such as a barrier, or an `Acquire`. The internal data structure should be a list of `instructions`. It should be possible to request the times of these instructions which would dynamically evaluate them or pull from a cached value if this had already been evaluated and not invalidated.

Reply:
barrier is not an instruction, acquire operates over time on AcquireChannel

Reply:
by channels is better, and possible if you think of it as hardware channels

Reply:
Yeah I think even if you have an instruction on multiple channels, you still can maintain a `_insts_by_chan` datastructure. You just repeat the instruction on both channels.

I mostly think about circuit schedules now, and in that case, given a 600 ns CNOT that engages 2 qubits, you would add a 600 ns Instruction to the timeline of both qubits.